### PR TITLE
Properly serialize key metadata in GenericDataFile.

### LINKED ISSUE #169: ByteBuffer encryption key metadata in GenericDataFile is not serializable
This will be problematic for the Spark data source in particular because `GenericDataFile` objects are serialized and sent back to the driver as task results.

### PR DESCRIPTION
**Pull request by mccheah**
Required since ByteBuffer itself isn't serializable. Override Java serialization to convert to byte array first before pushing to object streams.

Closes #169

### CODE REVIEW

**Comment by mccheah** [comment]
@rdsr or @rdblue care to review?

**Comment by rdblue** [comment]
The problem with using `byte[]` internally is that it would require copying the buffer when it is set. That copy would slow down scan planning.

File: core/src/main/java/org/apache/iceberg/GenericDataFile.java [hunk 1]
```diff
@@ -439,4 +442,29 @@ public String toString() {
     }
     return null;
   }
+
+  private void writeObject(ObjectOutputStream output) throws IOException {
```
Comment (rdblue) [inline core/src/main/java/org/apache/iceberg/GenericDataFile.java]: This looks correct to me, but I'm wondering if we should serialize to a `byte[]` in Avro's binary format. That would cut down on the custom serialization code. Probably not worth doing right now, but we should keep it in mind for later.

**Comment by rdblue** [comment]
Looks good to me. I'm going to merge this.

**Pull request merged**